torch/sparse/_triton_ops.py

```diff
@@ -563,30 +564,49 @@ def scatter_mm_meta(M, K, N, Ms, Ks,
                 num_stages=num_stages, num_warps=num_warps, SPLIT_N=SPLIT_N, **extra)
 
 
-def bsr_scatter_mm_indices_data(bsr, other, indices_format='bsr_strided_mm_compressed', **meta_input):
-    """Computes indices data for :func:`scatter_mm` used in BSR and
-    strided tensor matrix multiplication.
+class TensorAsKey:
```

Comment:
cc @albanD 

I think this might not always work?

Reply:
It depends how do you define "not working". The approach is not exact in the sense that otherwise equal tensors may represent different keys but that does not break tensor caching.
On the other hand, it is better than all tensors representing different keys, including tensors that share the storage and have equal strides which would make tensor caching rather pointless.

Reply:
There's (obviously) two key risks here

1. Correctness: Two Tensors are seen as equal, but they're actually not.
2. Performance: A lot of Tensors are being kept alive that would have otherwise been deallocated causing what maybe appear like an OOM (or making the use of sparsity cause worse memory footprint than not using sparsity).

I'm mostly worried about Correctness for now. 

I think the key you implemented here is comprehensive, but @albanD tends to have a complete view of the data structure. I want to make sure we're not missing anything here.

Reply:
I'm very confused by this. How do you know the content of these Tensors hasn't changed?

Reply:
yes, you are right in general: two tensors having the same data pointer and strides may be unequal if these have different, say, conjugate bits.

In this particular case, we consider only integer tensors that are returned by `crow/col/ccol/row_indices` methods. So, I think we are relatively safe here. To make sure that TensorAsKey is used to wrap only integer tensors, we can introduce dtype check in its constructor. Are there any tensor bits specific to integer tensors?

Re performance: to avoid unnecessary memory consumption, we can reduce the LRU cache size. Currently, we are using the default size 128.

Reply:
> How do you know the content of these Tensors hasn't changed?

@albanD As said above, the feature is used only for integer tensors returned by `crow/col/ccol/row_indices` methods. Such tensors cannot be changed in place (well, nothing prevents changing the content of a tensor but in the case of indices of sparse tensors, an arbitrary change of such tensors will likely cause a crash).

Reply:
If you're happy with ignoring this kind of silent correctness.
I would:
- Update the comment to make it clearer that we want to match any Tensor that looks at the same memory even if they're not the same Tensor object. (object sameness doesn't use equal, no need to mention it here).
- Add storage_offset to your key.
- You should find a way to use weakref to avoid holding onto the Tensor directly.

Reply:
Let's do this as a follow up PR.

Reply:
@albanD @cpuhrsch here's the follow-up PR: https://github.com/pytorch/pytorch/pull/112076